### PR TITLE
[IN-279][OSF] Add preprint retractions fields 

### PR DESCRIPTION
**Pull request by fabmiz**
## Purpose

Add retraction-related fields to preprint model

## Changes
- Add tests
- Add `HideIfNotWithdrawal` to allow hide retraction specific fields
- Add `date_retracted `, `retraction_justification `, `ever_public ` to preprint model
- Update api-exposed fields with `HideIfNotWithdrawal`, `HideIfWithdrawal` 

## Ticket

[[IN-279]](https://openscience.atlassian.net/browse/IN-279)

### CODE REVIEW

File: api/preprints/serializers.py [hunk 1]
```diff
@@ -73,71 +73,73 @@ class PreprintSerializer(TaxonomizableSerializerMixin, JSONAPISerializer):
     id = IDField(source='_id', read_only=True)
     type = TypeField()
 
-    date_created = VersionedDateTimeField(source='created', read_only=True)
+    date_created = VersionedDateTimeField(source='created', allow_null=True, read_only=True)
```
Comment (mfraezz) [inline api/preprints/serializers.py]: Why was this changed?

File: api/preprints/serializers.py [hunk 2]
```diff
@@ -73,71 +73,73 @@ class PreprintSerializer(TaxonomizableSerializerMixin, JSONAPISerializer):
...
-    is_published = ser.BooleanField(required=False)
-    is_preprint_orphan = ser.BooleanField(read_only=True)
-    license_record = NodeLicenseSerializer(required=False, source='license')
     title = ser.CharField(source='node.title', required=False)
     description = ser.CharField(required=False, allow_blank=True, allow_null=True, source='node.description')
-    tags = JSONAPIListField(child=NodeTagField(), required=False, source='node.tags')
-    node_is_public = ser.BooleanField(read_only=True, source='node__is_public')
-    preprint_doi_created = VersionedDateTimeField(read_only=True)
+    date_retracted = VersionedDateTimeField(read_only=True, allow_null=True)
+    retraction_justification = HideIfNotWithdrawal(ser.CharField(required=False, allow_blank=True))
```
Comment (mfraezz) [inline api/preprints/serializers.py]: It shouldn't affect anything, but for posterity/correctness this should also be `read_only`

File: api_tests/providers/preprints/serializers/test_serializers.py [hunk 3]
```diff
@@ -1,12 +1,59 @@
...
+
+        # test_non_retracted
+        req = make_drf_request_with_version()
+        result = PreprintSerializer(
+            self.preprint,
+            context={'request': req}).data
+        data = result['data']
+        attributes = set(data['attributes'])
+
+        assert_false(self.preprint.is_retracted)
```
Comment (mfraezz) [inline api_tests/providers/preprints/serializers/test_serializers.py]: Not necessarily a blocker, but in general when writing new tests/test classes, prefer `pytest.fixture`s and plain `assert`s over `nose`-style `setUp`/`tearDown` and `assert_something`s.

Comment (mfraezz) [reply api_tests/providers/preprints/serializers/test_serializers.py]: Please also test the behavior of `ever_public`

File: api/preprints/serializers.py (hunk 1 above)
Comment (fabmiz) [reply api/preprints/serializers.py]: probably confused`date_created`, `date_retracted` when adding `allow_null`. 😢

File: api/preprints/serializers.py [hunk 4]
```diff
@@ -78,66 +78,68 @@ class PreprintSerializer(TaxonomizableSerializerMixin, JSONAPISerializer):
...
     description = ser.CharField(required=False, allow_blank=True, allow_null=True, source='node.description')
-    tags = JSONAPIListField(child=NodeTagField(), required=False, source='node.tags')
-    node_is_public = ser.BooleanField(read_only=True, source='node__is_public')
-    preprint_doi_created = VersionedDateTimeField(read_only=True)
+    date_retracted = VersionedDateTimeField(read_only=True, allow_null=True)
+    retraction_justification = HideIfNotWithdrawal(ser.CharField(required=False, read_only=True, allow_blank=True))
+    is_published = HideIfWithdrawal(ser.BooleanField(required=False))
+    is_preprint_orphan = HideIfWithdrawal(ser.BooleanField(read_only=True))
+    license_record = HideIfWithdrawal(NodeLicenseSerializer(required=False, source='license'))
+    tags = HideIfWithdrawal(JSONAPIListField(child=NodeTagField(), required=False, source='node.tags'))
```
[the remark this answers is not in the document]
Comment (mfraezz) [reply api/preprints/serializers.py]: Whoops, nice catch ⚾️
